homeassistant/components/zwave_js/discovery.py

```diff
@@ -0,0 +1,138 @@
+    # [optional] the value's endpoint must match ANY of these values
+    endpoint: List[int] = field(default_factory=list)
+    # [optional] the value's property must match ANY of these values
+    property: List[str] = field(default_factory=list)
+    # [optional] the value's metadata_type must match ANY of these values
+    type: List[str] = field(default_factory=list)
+
+
+DISCOVERY_SCHEMAS = [
+    # generic text sensors
+    ZWaveDiscoverySchema(
+        platform="sensor",
+        hint="string_sensor",
+        command_class=[
+            CommandClass.ALARM,
+            CommandClass.SENSOR_ALARM,
+            CommandClass.INDICATOR,
+            CommandClass.NOTIFICATION,
+        ],
+        type=["string"],
+    ),
+    # generic numeric sensors
+    ZWaveDiscoverySchema(
+        platform="sensor",
+        hint="numeric_sensor",
+        command_class=[
+            CommandClass.SENSOR_MULTILEVEL,
+            CommandClass.METER,
+            CommandClass.ALARM,
+            CommandClass.SENSOR_ALARM,
+            CommandClass.INDICATOR,
+            CommandClass.BATTERY,
+            CommandClass.NOTIFICATION,
+            CommandClass.BASIC,
+        ],
+        type=["number"],
+    ),
+]
+
+
+@callback
+def async_discover_values(node: ZwaveNode) -> AsyncGenerator[ZwaveDiscoveryInfo, None]:
+    """Run discovery on ZWave node and return matching (primary) values."""
+    for value in node.values.values():
+        disc_val = async_discover_value(value)
+        if disc_val:
+            yield disc_val
+
+
+@callback
+def async_discover_value(value: ZwaveValue) -> Optional[ZwaveDiscoveryInfo]:
+    """Run discovery on Z-Wave value and return ZwaveDiscoveryInfo if match found."""
+    for schema in DISCOVERY_SCHEMAS:
+        # check device_class_basic
+        if not compare_value(schema.device_class_basic, value.node.device_class.basic):
+            continue
+        # check device_class_generic
+        if not compare_value(
+            schema.device_class_generic, value.node.device_class.generic
+        ):
+            continue
+        # check device_class_specific
+        if not compare_value(
+            schema.device_class_specific, value.node.device_class.specific
+        ):
+            continue
+        # check command_class
+        if not compare_value(schema.command_class, value.command_class):
+            continue
+        # check endpoint
+        if not compare_value(schema.endpoint, value.endpoint):
+            continue
+        # check property
+        if not compare_value(schema.property, value.property_):
+            continue
+        # check metadata_type
+        if not compare_value(schema.type, value.metadata.type):
+            continue
+        # all checks passed, this value belongs to an entity
+        return ZwaveDiscoveryInfo(
+            node=value.node,
+            primary_value=value,
+            platform=schema.platform,
+            platform_hint=schema.hint,
+        )
+
```

Comment:
```suggestion

        return None
```

Reply:
isn't None the default ? why explicitly return it ? For code readability ?

Reply:
Yes. We should always return a value explicitly at all exits if we return a value anywhere in a function.